README.md
# gen.parRep (v1.2)

C++ implementation of the Generalized Parallel Replica algorithm

Link to this repository : 

* https://gitlab.inria.fr/parallel-replica/gen.parRep

Link to the project gitlab page (includes more resources):

* https://gitlab.inria.fr/parallel-replica

See the following articles:

* A generalized parallel replica dynamics: Binder, Lelièvre & Simpson, 2015: https://doi.org/10.1016/j.jcp.2015.01.002

* gen.parRep: a first implementation of the Generalized Parallel Replica dynamics for the long time simulation of metastable biochemical systems: Hédin & Lelièvre, 2018: https://arxiv.org/abs/1807.02431 and https://doi.org/10.1016/j.cpc.2019.01.005

Molecular dynamics is performed by using external codes linked to this program such as:

* OpenMM (https://github.com/pandegroup/openmm and https://simtk.org/home/openmm)

If you find any error, bug, or limitation, please open an issue on https://gitlab.inria.fr/parallel-replica/gen.parRep/issues so that the community can benefit from its correction ; also feel free to clone/fork the repo and perform the modifications yourself if you have the programming abilities to do so !

Continuous Integration is performed on a test case system (see ./ci), ensuring that no release breaks the ability to compile the software; furthermore reference results from a previous run are available in ./ci/ref.sim in an attempt to achieve reproducibility whenever possible.

[![pipeline status](https://gitlab.inria.fr/parallel-replica/gen.parRep/badges/master/pipeline.svg)](https://gitlab.inria.fr/parallel-replica/gen.parRep/commits/master)

## DOCUMENTATION

Please consult the Gitlab wiki for an overview of this software use: 

https://gitlab.inria.fr/parallel-replica/gen.parRep/wikis/home

Doxygen can be used for generating a code reference documentation within the docs directory (particularly useful if you plan to modify/extend the software):

* doxygen parRep.doxy 

By default an HTML reference is generated (ready to read), and a docs/latex directory is also created, although the pdf version requires to be manually compiled (pdfLaTeX, see the Makefile in ./docs/latex).

The Lua input files, in ./mol, used for starting simulations, are self documented, and together
with the bash submission files in the ./run directory, it should be enough for starting to use the program.

## DEPENDENCIES

### Dependencies required before running the CMakeLists.txt script

You will need an MPI development framework installed, compatible with the MPI 3.0 or newer standard : tested implementations :

* Open MPI 1.10.2 (from Ubuntu 16.04 repositories)

CMake will try to locate it automatically.

You will need the SQLite3 dynamic library on your system. See :

* https://sqlite.org/index.html

### Dependencies downloaded automatically via the script "build/download_deps_and_run_cmake.sh"

The script "build/download_deps_and_run_cmake.sh" will download and compile if required the OpenMM and LuaJIT Dependencies.

You can however use your own version of the OpenMM library, or any Lua implementation (either compiled yourself or downloaded somewhere).

#### Ressources (OpenMM): 

* see https://simtk.org/home/openmm

* and/or https://github.com/pandegroup/openmm

* tested with version 7.0 to 7.3.

You may need NVIDIA CUDA or AMD OpenCL toolkit for enabling GPU acceleration ; see OpenMM documentation.
You may also need to edit CMakeLists.txt for specifying path to the include and lib directories of OpenMM.
CMake will try to locate it automatically.

#### Ressources (Lua/LuaJIT): 

A Lua implementation compatible with Lua API version >= 5.1 is required :

* A release of Lua 5.x is usually already installed by default for recent linux versions, try to execute 'lua' and/or 'locate liblua'

* You can Download and compile the official implementation : http://www.lua.org/download.html

* The LuaJIT implementation can provide an important speedup : http://luajit.org/download.html

Please download and compile any of the two. See below for hints for setting DCMAKE_PREFIX_PATH in case of a manual install.
CMake will try to locate it automatically.

### Dependencies provided in directory "external"

The excellent Sol2 (header-only Lua<->c++ interface) and LuaSQLite3 (a Lua/SQLite3 binding) Dependencies are provided
in the './external' directory and are automatically included/compiled if required, and thus should not require extra
configuration. See :

* https://github.com/ThePhD/sol2

* http://lua.sqlite.org/index.cgi/home

## COMPILE & INSTALL

C and C++ compiler compatible with the C99 and C++14 standards are required.
Tested compilers:

* gcc/g++ 5.4.0 and 6.2.0 (from Ubuntu 16.04 repositories)

Be sure to have CMake installed (http://www.cmake.org/), available on most repositories.
Tested version :

* CMake 3.5.1 (from Ubuntu 16.04 repositories)

Create a build directory and move to that directory:

* mkdir build && cd build

For building a debug or release or an intermediate release with debug information, do:

* cmake -DCMAKE_BUILD_TYPE=Debug ..

* cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo ..

* cmake -DCMAKE_BUILD_TYPE=Release ..         (default, with cpu targeting for the current compilation machine)

Debug builds are possibly slower, usually more memory consuming, but useful when debugging with gdb or Valgrind.

If some dependencies were note detected (e.g. because they are not available within /usr/local)
it is required to add the path where the were installed to CMAKE_PREFIX_PATH, e.g.:

* cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH="$HOME/bin/something" ..

Then, once CMake built a Makefile without error, just execute the following in order to build the executable:

* make

For a verbose make, use:

* make VERBOSE=1

For a faster multi-core (on N cores) build, use:

* make -j N

Please never edit the automatically generated Makefile, edit the CMakeLists.txt instead.

For specifying another compiler on linux, for example clang, or a proprietary one like Intel icc (untested) :

* CC=clang CXX=clang++ cmake ..

* CC=icc CXX=icpc cmake ..

In any case, you may need to edit the variables CMAKE_C_FLAGS_* and CMAKE_CXX_FLAGS_*
for setting proper levels of optimization. By default 

When using OpenMPI and forcing another C/C++ compilers version, the following might be required (possibly unsafe !):

* OMPI_CC=gcc-5 OMPI_CXX=g++-5 CC=gcc-5 CXX=g++-5 ccmake ..

## LICENSING (all files excepted subdirectory external and its content)

Copyright (c) 2016-2019, Florent Hédin, Tony Lelièvre, and École des Ponts - ParisTech
All rights reserved.

The 3-clause BSD license is applied to this software.

See LICENSE.txt

See in ./external/* for licensing information concerning the embedded libraries provided in './external' 

## NOTES CONCERNING OpenMM

The software should automatically detect the fastest OpenMM Platform available on your computer (i.e. CUDA, OpenCL, ...)
If not it will run with the slow Reference platform : it is most probably because the OpenMM directory with the
'plugins' library is not found.

You may need to export the following OPENMM_PLUGIN_DIR environment variable to solve the problem : 

For example for a custom installation in /home/$USER/bin/openmm

* export OPENMM_PLUGIN_DIR=$HOME/bin/openmm/lib/plugins

For setting the amount of cpu threads used by each MPI process use the following env. variable :

* export OPENMM_CPU_THREADS=1

## Directory structure

**src** and **include** -> contains the C++ source and headers of the gen.parRep software.

**build** -> directory where dependencies will be downloaded, and the program built.

**external** -> contains source and headers of dependencies used by the gen.parRep software.

**cmake** -> contains files required for locating external packages (e.g. OpenMM and a MPI implementation).

**docs** -> contains files required for building the documentation using Doxygen.

**mol** -> contains ready to use test systems, i.e. input files and OpenMM configurations of various molecular systems.

**run** -> contains bash scripts demonstrating how to run the software using either mpirun or the SLURM scheduler; also contains bash files for running the software under the following debuggers/profilers: GDB, Valgrind and Scalasca.

**ci** -> contains a molecular system, an input script and reference data for performing continuous integration and non regression tests.


CHANGELOG.md
# v1.2.0

Mar 08th, 2019

The software was modified in order to allow reproducibility in some cases (the limiting factor is OpenMM which does not
always provides deterministic output even when using the same seeds (!), see http://docs.openmm.org/latest/userguide/library.html#determinism ).

The main executable now has 2 more command line options, '--inp-seeds [fname]' or '--out-seeds [fname]' for respectively loading seeds or writting seeds 
from/to a unique binary file. See rand.hpp/rand.cpp, od the Doxygen doc for more details.

These modifications now allow Continuous Integration (CI) on the infrastructure provided by INRIA : in mol/ci a small test case will be executed at each commit to the repository and compared to reference results.

The two other minor modifications concern the Lua scripts:

* "get_minimised_energy_crdvels" was added to the set of functions that the user can call from the Lua script : it simply combines in one call what 
"get_minimised_energy" and "get_minimised_crdvels" already provided.

* a extra optional parameter is available for the "simulation" parameters block when "simulation.algorithm" is "PARREP_FV" ; this parameter is
"simulation.minAccumulatedObs" : it will enforce that at least minAccumulatedObs observations of an observable have already been accumulated
before the convergence test is performed ; this may be useful if there is a risk of early pseudo-convergence for some of the observables when only a few samples have been accumulated.


**Download sources:**

https://gitlab.inria.fr/parallel-replica/gen.parRep/tags/v1.2.0

or

https://github.com/FHedin/gen.parRep/releases/tag/v1.2.0

----------------------------------------------

# v1.1.0

Nov 16th, 2018

Finished the implementation of a "MD_interface" abstract class for future addition of different Molecular Dynamics engines.
This required important changes to a significant part of the code.
Optimizations to the MPI communications were also performed, therefore the minor version number
of the software was incremented.

**Download sources:**

https://gitlab.inria.fr/parallel-replica/gen.parRep/tags/v1.1.0

or

https://github.com/FHedin/gen.parRep/releases/tag/v1.1.0

----------------------------------------------

# v1.0.1

Aug 16th, 2018

Simply added directory structure to readme file.

**Download sources:**

https://gitlab.inria.fr/parallel-replica/gen.parRep/tags/v1.0.1

or

https://github.com/FHedin/gen.parRep/releases/tag/v1.0.1

----------------------------------------------

# v1.0.0

Aug 13th, 2018

Release of v1.0.0 : several bug fixes like memory leaks addressed, added input files for alanine
dipeptide using transient propagation.

**Download sources:**

https://gitlab.inria.fr/parallel-replica/gen.parRep/tags/v1.0.0

or

https://github.com/FHedin/gen.parRep/releases/tag/v1.0.0

----------------------------------------------

# 1.0.0-rc1

July 6th, 2018

First public release of the software.
It is following the submission of the corresponding article on arXiv ; this is a release candidate 1.0.0-rc1,
the 1.0.0 release should follow within the next weeks.

**Download sources:**

https://gitlab.inria.fr/parallel-replica/gen.parRep/tags/1.0.0-rc1

or

https://github.com/FHedin/gen.parRep/releases/tag/1.0.0-rc1


src/main.cpp
/**
 * \file main.cpp
 * \copyright The 3-clause BSD license is applied to this software.
 * \author Florent Hédin
 * \author Tony Lelièvre
 * \author École des Ponts - ParisTech
 * \date 2016-2019
 */

#include <string>

#include <cstdlib>
#include <cstdio>
#include <cstring>

#include "global.hpp"
#include "logger.hpp"
#include "runSim.hpp"

#include "mpi_utils.hpp"

using namespace std;

// -----------------------------------------------------------------------------------------

/*
 * Some global variables initialisation here
 */

/*
 * Errors, warning, etc ... --> logging.
 * Default Level is LOG_WARNING, which means that everything which is at least
 * a warning is printed (it includes Errors also).
 * See logger.h for other possibilities.
 */
LOG_LEVELS LOG_SEVERITY = LOG_WARNING;

/*
 *  End of global variables initialisation
 */

// -----------------------------------------------------------------------------------------

//prototypes of functions in this c++ unit
void help(char *argv[]);

// -----------------------------------------------------------------------------------------
/**
 * \brief   Main entry of the program
 * \details The 2 variables \b argc and \b argv are used for extracting
 *          command line parameters.
 * \param   argc Number of arguments, at least one as \b argv contains at least the program's name.
 * \param   argv Array of character strings containing the arguments.
 * \return  On exit returns EXIT_SUCCESS, EXIT_FAILURE otherwise.
 */
int main(int argc, char* argv[])
{
  int32_t version,subversion;
  MPI_Get_version(&version,&subversion);
  
  char  mpi_lib_version_string[MPI_MAX_LIBRARY_VERSION_STRING];
  int32_t  str_len = MPI_MAX_LIBRARY_VERSION_STRING;
  MPI_Get_library_version(mpi_lib_version_string,&str_len);

  if(version < 3)
  {
    fprintf(stderr,"[ERROR] Your MPI runtime version was found to be %d.%d, however version 3.0 or newer is required !\n",version,subversion);
    
    fprintf(stderr,"[NOTE] If you have compiled this program yourself, you have a valid 3.0 (or newer) runtime available somewhere because this was already checked at compile time: therefore you are probably facing a dynamic library mismatch !! Please check the following environment variables:\n");
    
    fprintf(stderr,"\t PATH : %s\n",getenv("PATH"));
    fprintf(stderr,"\t LIBRARY_PATH : %s\n",getenv("LIBRARY_PATH"));
    fprintf(stderr,"\t LD_LIBRARY_PATH : %s\n",getenv("LD_LIBRARY_PATH"));
    
    return EXIT_FAILURE;
  }
  
  /*
   * This programm allows multiple threads per MPI rank for the MD engine, however only the thread who called
   * MPI_Init_thread here will perform MPI calls ; this implies that the MD engine SHALL NOT do any MPI call !!
   */
  int threads_request_level = MPI_THREAD_FUNNELED;
  int library_provided_threads_support;
  MPI_Init_thread(&argc, &argv, threads_request_level, &library_provided_threads_support);
  if(library_provided_threads_support != threads_request_level)
  {
    // If the MPI runtime does not support MPI_THREAD_FUNNELED (it should always be available but who knows...) : error
    fprintf(stderr,"[ERROR] Your MPI runtime apparently doesn't support MPI_Init_thread called with the following theads support : MPI_THREAD_FUNNELED! \n");
    MPI_Barrier(MPI_COMM_WORLD);
    MPI_CUSTOM_ABORT_MACRO();
  }
  
  // be sure that when exiting log files are properly flushed
  atexit(LOG_FLUSH_ALL);

  MPI_Barrier(MPI_COMM_WORLD);

  /* find out MY process ID, and how many processes were started. */
  int32_t my_id     = -1;
  int32_t num_procs = -1;
  MPI_Comm_rank(MPI_COMM_WORLD, &my_id);
  MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
  
  MPI_Barrier(MPI_COMM_WORLD);

  /* arguments parsing, we need at least "prog_name -i an_input_file"
   * prints some more instructions if needed
   */
  if(argc < 3)
  {
    fprintf(stdout,"[Info] No input file ! \n");
    help(argv);
    
    MPI_CUSTOM_ABORT_MACRO();
  }
  MPI_Barrier(MPI_COMM_WORLD);
  
  string input_file = NULLFILE;
  string seeds_inp_file = NULLFILE;
  string seeds_out_file = NULLFILE;

  // arguments parsing
  for (uint32_t i=1; i<(uint32_t)argc; i++)
  {
    // get name of input file
    if(!strcasecmp(argv[i],"-i"))
    {
      input_file = string(argv[++i]);
    }
    /*
     * reopen stdout to unique file based on the user specified template name ;
     * expects one dot and an extension at the end (exception: redirection to /dev/null is allowed)
     */
    else if(!strcasecmp(argv[i],"-o"))
    {
      string my_name(argv[++i]);
      MPIutils::mpi_get_unique_name(my_name);
      stdout = freopen(my_name.c_str(),"w",stdout);
    }
    /*
     * reopen stderr to unique file based on the user specified template name ;
     * expects one dot and an extension at the end (exception: redirection to /dev/null is allowed)
     */
    else if(!strcasecmp(argv[i],"-e"))
    {
      string my_name(argv[++i]);
      MPIutils::mpi_get_unique_name(my_name);
      stderr = freopen(my_name.c_str(),"w",stderr);
    }
    // specify the logging level
    else if(!strcasecmp(argv[i],"-log"))
    {
      if(!strcasecmp(argv[++i],"no"))
      {
        LOG_SEVERITY = LOG_NOTHING;
      }
      else if(!strcasecmp(argv[i],"warn"))
      {
        LOG_SEVERITY = LOG_WARNING;
      }
      else if(!strcasecmp(argv[i],"info"))
      {
        LOG_SEVERITY = LOG_INFO;
      }
      else if(!strcasecmp(argv[i],"dbg"))
      {
        LOG_SEVERITY = LOG_DEBUG;
      }
      else
        fprintf(stdout,"[Warning] Unknown log level '%s' : default value used.\n\n",argv[i]);
    }
    // in or out files where random seeds are read/written
    else if( !strcasecmp(argv[i],"--inp-seeds") )
    {
      seeds_inp_file = string(argv[++i]);
    }
    else if( !strcasecmp(argv[i],"--out-seeds") )
    {
      seeds_out_file = string(argv[++i]);
    }
    // print help and proper exit
    else if( !strcasecmp(argv[i],"-h") || !strcasecmp(argv[i],"-help") || !strcasecmp(argv[i],"--help") )
    {
      help(argv);
      MPI_CUSTOM_ABORT_MACRO();
    }
    // error if unknown command line option
    else
    {
      fprintf(stdout,"[Error] Argument '%s' is unknown.\n",argv[i]);
      help(argv);
      MPI_CUSTOM_ABORT_MACRO();
    }
  }

  // wait for parsing of command line on each rank
  MPI_Barrier(MPI_COMM_WORLD);
  
  //prepare log files if necessary
  try
  {
    init_logfiles();
  }
  catch(exception& e)
  {
    fprintf(stderr,"std::exception captured when initializing log files ! Error message is %s \n\n",e.what());
    MPI_CUSTOM_ABORT_MACRO();
  }
  MPI_Barrier(MPI_COMM_WORLD);
  
  if(my_id==0)
  {
    // Print date and some env. variables
    fprintf(stdout,"Welcome to %s ! Command line arguments succesfully parsed, now initialising...\n\n",argv[0]);
    fprintf(stdout,"Logging level is : %s : see the documentation to see which .log files are generated, and what they contain.\n\n",get_loglevel_string());
    fprintf(stdout,"DATE : %s\n",get_time());
    fprintf(stdout,"USER : %s\n",getenv("USER"));
    fprintf(stdout,"PWD : %s\n",getenv("PWD"));
  }

  // run the simulation
  run_simulation(input_file,seeds_inp_file,seeds_out_file);

  // closing log files is the last thing to do as errors may occur at the end
  close_logfiles();
  
  fprintf(stdout,"End of program\n");
  
  fflush(stdout);
  fflush(stderr);

  MPI_Finalize();
  
  return EXIT_SUCCESS;
}

// -----------------------------------------------------------------------------------------
/**
 * \brief   This function simply prints a basic help message.
 *
 * \details If any of \b -h or \b -help or \b --help are provided on the command line this help message is printed.\n
 *          If no command line parameter is present this message is also printed.\n
 *          If an unknown command line parameter is present this message is also printed.
 *
 * \param   argv Simply the same array of command line parameters, from function \b #main.
 */
void help(char *argv[])
{
  fprintf(stdout,"Need at least one argument : %s -i an_input_file\n",argv[0]);
  fprintf(stdout,"optional args : -o [stdout_to_file] -e [stderr_to_file] -log [logging level, one of { no | warn | info | dbg }] --inp-seeds [seed_file] --out-seeds [seed_file]\n");
  fprintf(stdout,"Examples : \n\t%s -i input_file -o out.txt -log info --inp-seeds seeds.bin \n",argv[0]);
  fprintf(stdout,"\t%s -i input_file -o out.txt -log info --out-seeds seeds.bin \n",argv[0]);
  fprintf(stdout,"The default logging level is 'warn' \n");
  fprintf(stdout,"See the documentation for more details. \n");
}

/*!
 * \mainpage A C++ implementation of the Generalized Parallel Replica algorithm
 *
 * Metastability is one of the major encountered obstacle when performing long molecular dynamics simulations,
 * and many methods were developed with the aim of addressing this challenge.
 * 
 * The ``Parallel Replica''(parRep) dynamics is known for allowing to simulate very long
 * trajectories of metastable Langevin dynamics in the materials science community,
 * but it relies on assumptions that can hardly be transposed to the world of biochemical simulations.
 *
 * The later developed ``Generalized parRep'' variant solves those issues, but it was not applied to
 * significant systems of interest so far.
 *
 * This software is a new implementation of the Generalized parallel replica method, targeting
 * frequently encountered metastable biochemical systems.
 * It was used in version v1.0.0 for producing all the simulations published in : 
 * 
 *  <a href="https://doi.org/10.1016/j.cpc.2019.01.005">https://doi.org/10.1016/j.cpc.2019.01.005</a>
 * 
 * \image html  main.png
 * \image latex main.pdf
 * 
 */
